### DISCUSSION TITLE
Nesting types implementing `De/SerializeAs` with `serde_as`

### DISCUSSION THREAD

**Discussion by spikespaz** [jonasbb/serde_with#480] [Q&A] [answered]
I am attempting to use `serde_as` to chain serialization and deserialization in a strange way. I have a hard time explaining it so I think code may make more sense than words here.

```rust
#[serde_as]
#[derive(Clone, Debug, Serialize, Deserialize)]
struct TestType {
    #[serde_as(as = "JsonString<Vec<Base62Encoded<u64>>>")]
    field: Vec<u64>,
}
```

The type `Base62Encoded` is defined as follows:

```rust
use std::str::FromStr;

use serde::{Deserialize, Deserializer, Serialize, Serializer};
use serde_with::{DeserializeAs, DeserializeFromStr, SerializeAs, SerializeDisplay};

#[derive(Clone, Debug, PartialEq, SerializeDisplay, DeserializeFromStr)]
pub struct Base62Encoded<T>(T);

impl<T> std::fmt::Display for Base62Encoded<T>
where
    T: Clone + Into<u128>,
{
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        formatter.write_str(&base62::encode(self.0.clone()))
    }
}

impl<T> FromStr for Base62Encoded<T>
where
    u128: TryInto<T>,
{
    type Err = base62::DecodeError;

    fn from_str(other: &str) -> Result<Self, Self::Err> {
        match base62::decode(other)?.try_into() {
            Ok(n) => Ok(Base62Encoded(n)),
            Err(_) => Err(Self::Err::ArithmeticOverflow),
        }
    }
}

impl<T> SerializeAs<T> for Base62Encoded<T>
where
    T: Clone + Into<u128>,
{
    fn serialize_as<S>(source: &T, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        Base62Encoded(source.clone()).serialize(serializer)
    }
}

impl<'de, T> DeserializeAs<'de, T> for Base62Encoded<T>
where
    u128: TryInto<T>,
{
    fn deserialize_as<D>(deserializer: D) -> Result<T, D::Error>
    where
        D: Deserializer<'de>,
    {
        Base62Encoded::deserialize(deserializer).map(|this| this.0)
    }
}
```

This does appear to do the job as expected. Now, this works fine if I use `#[serde_as(as = "Vec<Base62Encoded<u64>>")]` on a field that is of the type `Vec<u64>`. I then re-implemented `JsonString` intending to wrap the type to get a JSON-encoded array of base-62-encoded integers:

```rust
use std::marker::PhantomData;

use serde::de::{DeserializeOwned, Error as DeserializeError, Visitor};
use serde::ser::Error as SerializeError;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use serde_with::{DeserializeAs, SerializeAs};

#[derive(Clone, Debug, PartialEq)]
pub struct JsonString<T>(T);

impl<T> Serialize for JsonString<T> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.collect_str(&serde_json::to_string(self).map_err(SerializeError::custom)?)
    }
}

impl<'de, T> Deserialize<'de> for JsonString<T>
where
    T: DeserializeOwned,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct JsonStringVisitor<T>(PhantomData<T>);

        impl<'de, T> Visitor<'de> for JsonStringVisitor<T>
        where
            T: DeserializeOwned,
        {
            type Value = JsonString<T>;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                formatter.write_str("a value that can be serialized as a JSON string")
            }

            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
            where
                E: DeserializeError,
            {
                Ok(JsonString(
                    serde_json::from_str(value).map_err(DeserializeError::custom)?,
                ))
            }
        }

        deserializer.deserialize_str(JsonStringVisitor(PhantomData))
    }
}

impl<T> SerializeAs<T> for JsonString<T>
where
    T: Clone,
{
    fn serialize_as<S>(source: &T, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        JsonString(source.clone()).serialize(serializer)
    }
}

impl<'de, T> DeserializeAs<'de, T> for JsonString<T>
where
    T: DeserializeOwned,
{
    fn deserialize_as<D>(deserializer: D) -> Result<T, D::Error>
    where
        D: Deserializer<'de>,
    {
        JsonString::deserialize(deserializer).map(|this| this.0)
    }
}
```

Using all of these in combination tells me that the compiler is expecting a totally different type:

```
error[E0308]: mismatched types
  --> src/endpoints.rs:16:24
   |
16 | #[derive(Clone, Debug, Serialize, Deserialize)]
   |                        ^^^^^^^^^ expected struct `base62::Base62Encoded`, found `u64`
   |
   = note: expected reference `&std::vec::Vec<base62::Base62Encoded<u64>>`
              found reference `&'__a std::vec::Vec<u64>`
   = note: this error originates in the derive macro `Serialize` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> src/endpoints.rs:16:35
   |
16 | #[derive(Clone, Debug, Serialize, Deserialize)]
   |                                   ^^^^^^^^^^^ expected `u64`, found struct `base62::Base62Encoded`
   |
   = note: expected struct `std::vec::Vec<u64>`
              found struct `std::vec::Vec<base62::Base62Encoded<u64>>`
   = note: this error originates in the macro `try` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I apologize for not asking a clearly defined question.

**Comment by jonasbb** [comment, marked as answer]
Hi, if you want to get into more complicated implementations, I encourage you to read the documentation of [`SerializeAs`](https://docs.rs/serde_with/1.14.0/serde_with/trait.SerializeAs.html). It contains some examples which show how to implement the code.

The problem is that you are asking `serde_json` to deserialize a `Vec<Base62Encoded<u64>>`, but you need to produce a `Vec<u64>`. You need to mark in your type signature, that the `T` inside `JsonString` is different than the `T` you return. You turn the `JsonString` into a container-like type.

I would be happy to accept a PR which changes the `JsonString` in this crate to allow for this added flexibility.

```rust
pub struct JsonString<T>(PhantomData<T>);

impl<T, TAs> SerializeAs<T> for JsonString<TAs>
where
    TAs: SerializeAs<T>,
{
    fn serialize_as<S>(source: &T, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        // `collect_str` will needlessly call `Display` again.
        // Since you already have a `String` `serialize_str` is better
        serializer.serialize_str(
            &serde_json::to_string(&serde_with::ser::SerializeAsWrap::<T, TAs>::new(source))
                .map_err(SerializeError::custom)?,
        )
    }
}
```

---

A couple of notes on your code:

> `JsonString(source.clone()).serialize(serializer)`

This looks like unbounded recursion to me. To use `SerializeAs` your types also do not need to implement `Serialize`.

The `Base62Encoded` is also quite complicated, going from `SerializeAs` to `Serialize` to `Display`. You can have a look here, for a similar type and see how it is implemented.
https://docs.rs/serde_with/1.14.0/serde_with/base64/struct.Base64.html

**Reply by spikespaz** [reply]
Hi Jonas, would you mind adding some documentation to the [`SerializeAsWrap`](https://docs.rs/serde_with/1.14.0/serde_with/ser/struct.SerializeAsWrap.html) type? I am having a hard time getting when it should or shouldn't be used, and am running into issues satisfying generic bounds that I don't think correct.

In particular, is a type considered a wrapper even if the type is not intended to be used directly, but instead only as a `SerializeAs` and `DeserializeAs` **adapter** type?

**Reply by jonasbb** [reply]
The documentation already links to `SerializeAs` and there it is explained in my opinion. You use it if you need something which implements `Serialize` but only have something which implements `SerializeAs`. See the `Box` example here <https://docs.rs/serde_with/latest/serde_with/ser/trait.SerializeAs.html#differences-to-serialize>.

If you have concrete suggestions on how to improve it, I am happy to hear them.

**Reply by spikespaz** [reply]
I see that documentation. I need some help understanding what the purpose of it is, because I see it used in the "inner" serialize in the example you gave, and I also see it used for `Box<T>` and `Box<U>`. Perhaps a description that covers more potential cases where `T` and `U` are known types would help me understand, because as it is, my brain is trying to compile the innards of your library, the source for which I have not read all of, and my trait implementations with unknown types. I do not have the same high-level overview that you have because I didn't design the library. I have a hard time satisfying generic bounds when I don't exactly know what the types are supposed to represent in each situation.

Perhaps giving you another sample of my code to fix can alleviate my struggles, as it did last time. In your `Base64` and `JsonString` examples that I've read through, you've duplicated code when it comes to the implementation of `Serialize` and `SerializeAs`, and while it is a very small amount of code, and my problem would be solved within seconds with a simple copy/paste, I would like to understand the structure of a proper implementation of the behavior that you and Marcin have defined.

Following is my nearly-working code. The issue is the trait bound of `SerializeAs<_>` with the elided type in the only unit test. I do not want to satisfy the trait bound, because I do not believe it to be correct, it looks like unbounded recursion.

```rust
use std::marker::PhantomData;

use serde::{Deserialize, Deserializer, Serialize, Serializer};
use serde_with::ser::SerializeAsWrap;
use serde_with::{DeserializeAs, SerializeAs};

pub mod with {
    use std::fmt;
    use std::marker::PhantomData;

    use serde::de::{Error as DeserializeError, Unexpected, Visitor};
    use serde::{Deserializer, Serializer};

    pub fn serialize<S, T>(value: &T, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
        T: Clone + Into<u128>,
    {
        serializer.serialize_str(&base62::encode(value.clone()))
    }

    pub fn deserialize<'de, D, T>(deserializer: D) -> Result<T, D::Error>
    where
        D: Deserializer<'de>,
        u128: TryInto<T>,
    {
        struct _Visitor<T>(PhantomData<T>);

        impl<'de, T> Visitor<'de> for _Visitor<T>
        where
            u128: TryInto<T>,
        {
            type Value = T;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("a value that can be converted from a base-62 encoded u128")
            }

            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
            where
                E: DeserializeError,
            {
                base62::decode(value)
                    .map_err(DeserializeError::custom)?
                    .try_into()
                    .map_err(|_| DeserializeError::invalid_value(Unexpected::Str(value), &self))
            }
        }

        deserializer.deserialize_str(_Visitor(PhantomData))
    }
}

pub struct Base62<T>(PhantomData<T>);

impl<T, U> SerializeAs<T> for Base62<U>
where
    T: Into<u128>,
    U: SerializeAs<T>,
{
    fn serialize_as<S>(source: &T, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        SerializeAsWrap::<T, U>::new(source).serialize(serializer)
    }
}

impl<'de, T> DeserializeAs<'de, T> for Base62<T>
where
    u128: TryInto<T>,
{
    fn deserialize_as<D>(deserializer: D) -> Result<T, D::Error>
    where
        D: Deserializer<'de>,
    {
        with::deserialize(deserializer)
    }
}

#[cfg(test)]
pub mod tests {
    use serde::{Deserialize, Serialize};
    use serde_with::serde_as;

    use super::Base62;

    #[serde_as]
    #[derive(Serialize, Deserialize)]
    struct TestType<T>
    where
        T: Into<u128>,
        u128: TryInto<T>,
    {
        #[serde_as(as = "Vec<Base62<T>>")]
        pub values: Vec<T>,
    }

    // Test both serializing and deserializing in one go
    #[test]
    fn test_roundtrip() {
        // Create a thousand large numbers, and add a zero for pedanticism
        let range = ((u128::MAX - 1000)..u128::MAX).chain([0]);
        // Encode those without the wrapper
        let values = range.clone().map(base62::encode);
        // Create an instance of the wrapper type with the same numbers
        let container = TestType {
            values: range.collect(),
        };
        // Serialize the value as a JSON string
        let serialized = serde_json::to_string(&container).unwrap();
        // Parse the string back into a `Value` type
        let parsed = serde_json::from_str::<serde_json::Value>(&serialized).unwrap();
        // Keep unwrapping to get the vector of encoded strings
        let parsed = parsed
            .as_object()
            .unwrap()
            .get("values")
            .unwrap()
            .as_array()
            .unwrap()
            .iter()
            .map(|v| v.as_str().unwrap());

        // Iterate over the manually encoded numbers and the ones encoded by the
        // wrapper's implementation of `SerializeAs` and `Serialize`
        for (expect, actual) in std::iter::zip(values, parsed) {
            assert_eq!(&expect, &actual);
        }
    }
}
```

**Reply by spikespaz** [reply]
At first I had tried something like `SerializeAsWrap::<T, Base62<T>>::new(source).serialize(serializer)` but obviously this was unbounded recursion.

**Reply by spikespaz** [reply]
I have it working by removing the use of `SerializeAsWrap` altogether, so I believe that I am misunderstanding the purpose. I guess what needs clarification is: my type, `Base62<PhantomData<T>>`, is what I would call a "wrapper" type. It is a struct, therefore a form of data, but exists solely to define the behavior for the context in which it is supposed to be used.

Here is the working portion:
```rust
use std::marker::PhantomData;

use serde::{Deserialize, Deserializer, Serialize, Serializer};
use serde_with::ser::SerializeAsWrap;
use serde_with::{DeserializeAs, SerializeAs};

pub mod with {
    use std::fmt;
    use std::marker::PhantomData;

    use serde::de::{Error as DeserializeError, Unexpected, Visitor};
    use serde::{Deserializer, Serializer};

    pub fn serialize<S, T>(value: &T, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
        T: Clone + Into<u128>,
    {
        serializer.serialize_str(&base62::encode(value.clone()))
    }

    pub fn deserialize<'de, D, T>(deserializer: D) -> Result<T, D::Error>
    where
        D: Deserializer<'de>,
        u128: TryInto<T>,
    {
        struct _Visitor<T>(PhantomData<T>);

        impl<'de, T> Visitor<'de> for _Visitor<T>
        where
            u128: TryInto<T>,
        {
            type Value = T;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("a value that can be converted from a base-62 encoded u128")
            }

            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
            where
                E: DeserializeError,
            {
                base62::decode(value)
                    .map_err(DeserializeError::custom)?
                    .try_into()
                    .map_err(|_| DeserializeError::invalid_value(Unexpected::Str(value), &self))
            }
        }

        deserializer.deserialize_str(_Visitor(PhantomData))
    }
}

pub struct Base62<T>(PhantomData<T>);

impl<T> SerializeAs<T> for Base62<T>
where
    T: Clone + Into<u128>,
    // U: SerializeAs<T>,
{
    fn serialize_as<S>(source: &T, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        with::serialize(source, serializer)
        // SerializeAsWrap::<T, U>::new(source).serialize(serializer)
    }
}

impl<'de, T> DeserializeAs<'de, T> for Base62<T>
where
    u128: TryInto<T>,
{
    fn deserialize_as<D>(deserializer: D) -> Result<T, D::Error>
    where
        D: Deserializer<'de>,
    {
        with::deserialize(deserializer)
    }
}

#[cfg(test)]
pub mod tests {
    use serde::{Deserialize, Serialize};
    use serde_with::serde_as;

    use super::Base62;

    #[serde_as]
    #[derive(Serialize, Deserialize)]
    struct TestType<T>
    where
        T: Clone + Into<u128>,
        u128: TryInto<T>,
    {
        #[serde_as(as = "Vec<Base62<T>>")]
        pub values: Vec<T>,
    }

    // Test both serializing and deserializing in one go
    #[test]
    fn test_roundtrip() {
        // Create a thousand large numbers, and add a zero for pedanticism
        let range = ((u128::MAX - 1000)..u128::MAX).chain([0]);
        // Encode those without the wrapper
        let values = range.clone().map(base62::encode);
        // Create an instance of the wrapper type with the same numbers
        let container = TestType {
            values: range.collect(),
        };
        // Serialize the value as a JSON string
        let serialized = serde_json::to_string(&container).unwrap();
        // Parse the string back into a `Value` type
        let parsed = serde_json::from_str::<serde_json::Value>(&serialized).unwrap();
        // Keep unwrapping to get the vector of encoded strings
        let parsed = parsed
            .as_object()
            .unwrap()
            .get("values")
            .unwrap()
            .as_array()
            .unwrap()
            .iter()
            .map(|v| v.as_str().unwrap());

        // Iterate over the manually encoded numbers and the ones encoded by the
        // wrapper's implementation of `SerializeAs` and `Serialize`
        for (expect, actual) in std::iter::zip(values, parsed) {
            assert_eq!(&expect, &actual);
        }
    }
}
```

Now, typically the type `T: Into<u128>, u128: TryFrom<T>` is going to be a primitive that implements `Copy`, but I would like to eliminate the use of the `Clone` bound in `with::serialize`. If I were to guess, there is no way to do this without modifying the encode function in the `base62` crate, or using the alternative function for a byte buffer?

**Reply by spikespaz** [reply]
Ignoring previous attempts, I am now further along.

I don't know how to properly investigate this type mismatch.
```sh
cargo test --package awaur --lib --all-features -- serde_with::json_string::tests::test_wrapping
```
```rust
error[E0308]: mismatched types
   --> src/serde_with/json_string.rs:194:33
    |
194 |             #[derive(Serialize, Deserialize)]
    |                                 ^^^^^^^^^^^ expected `u64`, found struct `base62::wrapper::Base62`
    |
    = note: expected struct `Vec<u64>`
               found struct `Vec<base62::wrapper::Base62<u64>>`
    = note: this error originates in the macro `try` (in Nightly builds, run with -Z macro-backtrace for more info)
```
This is in the integration test that I wrote to see if wrapping works. It does not.
<https://github.com/spikespaz/awaur/blob/master/src/serde_with/json_string.rs#L194>

**Reply by jonasbb** [reply]
You use a different implementation for `JsonString` than I posted in my [original reply](https://github.com/jonasbb/serde_with/discussions/480#discussioncomment-2954804). Any reason you choose not to use this working implementation?

**Reply by spikespaz** [reply]
Because I am failing to understand why/how `SerializeAsWrap` and `DeserializeAsWrap` work, and am trying to do it without just to experiment. I know that my current code is wrong, but I do have `impl<T, U> SerializeAs<T> for JsonString<U>`. Is `U` inferred to be `u64`, or is it inferred to be `Base62<u64>`? I don't know what the compiler talks about when it says "expected" and "found".

> The trait is analogue to the [`serde::Serialize`](https://docs.rs/serde/1.0.137/serde/ser/trait.Serialize.html) trait, with the same meaning of input and output arguments. **It can and should the implemented using the same code structure as the [`Serialize`](https://docs.rs/serde/1.0.137/serde/ser/trait.Serialize.html) trait.** As such, the same advice for [implementing `Serialize`](https://serde.rs/impl-serialize.html) applies here.

The bold text is also my reason for wanting to use the exact same implementation (`awaur::serde_with::json_string::with::{serialize, deserialize}`) for the inner type `T`. I think I'm getting confused because `T` in `Serialize`/`SerializeAs` is a different type from `T` in `Deserialize/DeserializeAs`. I wanted to use `awaur::serde_with::{json_string, base62}::with::deserialize` so that the same `Visitor` is used.

> Any reason you choose not to use this working implementation?

I don't copy/paste code unless I totally understand it, so I'm still messing around with it.

**Reply by spikespaz** [reply]
`SerializeAsWrap` requires that `U: SerializeAs<T>` meaning that types that only implement `Serialize` do not work as inner types of `JsonString<U>`.

**Reply by jonasbb** [reply]
> > The trait is analogue to the [`serde::Serialize`](https://docs.rs/serde/1.0.137/serde/ser/trait.Serialize.html) trait, with the same meaning of input and output arguments. **It can and should the implemented using the same code structure as the [`Serialize`](https://docs.rs/serde/1.0.137/serde/ser/trait.Serialize.html) trait.** As such, the same advice for [implementing `Serialize`](https://serde.rs/impl-serialize.html) applies here.
> 
> The bold text is also my reason for wanting to use the exact same implementation (`awaur::serde_with::json_string::with::{serialize, deserialize}`) for the inner type `T`. [...]

The quoted code does not permit using the exact same implementation. It says the same **structure** will work in both cases. And that remains true. You will need to do adjustments like using `SerializeAs` instead of `Serialize`, after all the two traits serve slightly different purposes. But overall, you can often very much mirror the implementation. For example, compare [`impl<T> Serialize for Option<T>`](https://docs.rs/serde/latest/src/serde/ser/impls.rs.html#99-113) with [`impl<T, U> SerializeAs<Option<T>> for Option<U>`](https://docs.rs/serde_with/latest/src/serde_with/ser/impls.rs.html#72-85). Maybe that example also shows better why `SerializeAsWrap` is needed.

> `SerializeAsWrap` requires that `U: SerializeAs<T>` meaning that types that only implement `Serialize` do not work as inner types of `JsonString<U>`.

That is indeed how Rusts trait bounds work. If you only use `SerializeAs` in the bound, you cannot use `Serialize`. I fail to see how this is a problem though.
